File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/domain/policies/CreditGiftCardPolicy.java
package it.giannibombelli.wsc2026.giftcard.domain.policies;

import it.giannibombelli.wsc2026.booking.domain.events.BookingResultEvents;
import it.giannibombelli.wsc2026.common.domain.model.Policy;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;


public final class CreditGiftCardPolicy implements Policy<BookingResultEvents, CreditGiftCard> {
    @Override
    public CreditGiftCard evaluate(BookingResultEvents event) {
        Require.requireArgument(event, "Booking event");
        return switch (event) {
            case BookingResultEvents.BookingConfirmed it -> new CreditGiftCard(it.giftCardId(), it.amount());
            case BookingResultEvents.BookingRefused it -> new CreditGiftCard(it.giftCardId(), it.amount());
            case BookingResultEvents.BookingRejected ignored -> null;
        };
    }
}

File: typescript/src/giftcard/application/policies/creditGiftCardPolicy.ts
// Policy che traduce i risultati della prenotazione in accrediti sulla gift card.
// Violazione cross-BC didattica: GiftCard non dovrebbe importare tipi da Booking.

import type { BookingResultEvent } from '@/booking/domain/events/bookingResultEvents.js';
import type { Policy } from '@/common/application/policy.js';
import { match } from 'ts-pattern';
import { CreditGiftCard, creditGiftCard } from '../../application/commands/creditGiftCard.js';

export class CreditGiftCardPolicy implements Policy<BookingResultEvent, CreditGiftCard> {
  evaluate(event: BookingResultEvent): CreditGiftCard | null {
    return match(event)
      .with(
        { kind: 'BookingConfirmed' },
        (confirmed) => creditGiftCard(confirmed.giftCardId, confirmed.amount)
      )
      .with(
        { kind: 'BookingRefused' },
        (refused) => creditGiftCard(refused.giftCardId, refused.amount)
      )
      .with({ kind: 'BookingRejected' }, () => null)
      .exhaustive();
  }
}
